File: icsopenapi/easyopen-sdk/sdk-csharp/EasyopenSDKCSharp/Client/OpenClient.cs
using System;
using System.Web;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections;
using System.Threading.Tasks;

using EasyopenSDKCSharp.Common;
using EasyopenSDKCSharp.Request;
using EasyopenSDKCSharp.Response;
using EasyopenSDKCSharp.Utility;


namespace EasyopenSDKCSharp.Client
{
    /// <summary>
    /// 客户端
    /// </summary>
    public class OpenClient
    {
        private const string AND = "&";
        private const string EQ = "=";

        private static OpenConfig DEFAULT_CONFIG = new OpenConfig();

        private const string ACCEPT_LANGUAGE = "Accept-Language";
        private const string AUTHORIZATION = "Authorization";
        private const string PREFIX_BEARER = "Bearer ";

        private OpenConfig openConfig;

        public OpenConfig OpenConfig
        {
            get { return openConfig; }
            set { openConfig = value; }
        }

        private OpenRequest openRequest;      
  
        public OpenRequest OpenRequest
        {
            get { return openRequest; }
            set { openRequest = value; }
        }

        private string url;
        private string appKey;

        public string AppKey
        {
            get { return appKey; }
            set { appKey = value; }
        }
        private string secret;


        public OpenClient(string url, string appKey, string secret) : this(url, appKey, secret, DEFAULT_CONFIG)
        {
            
        }

        public OpenClient(string url, string appKey, string secret, OpenConfig openConfig)
        {
            this.url = url;
            this.appKey = appKey;
            this.secret = secret;
            this.openConfig = openConfig;
            this.openRequest = new OpenRequest(openConfig);
        }

        /// <summary>
        /// 发送请求
        /// </summary>
        /// <typeparam name="T">返回的Response类</typeparam>
        /// <param name="request">请求对象</param>
        /// <returns>返回Response类</returns>
        public virtual T Execute<T>(BaseRequest<T> request)
        {
            return this.Execute<T>(request, null);
        }

        /// <summary>
        /// 发送请求
        /// </summary>
        /// <typeparam name="T">返回的Response类</typeparam>
        /// <param name="request">请求对象</param>
        /// <param name="jwt">jwt字符串</param>
        /// <returns>返回Response类</returns>
        public virtual T Execute<T>(BaseRequest<T> request, string jwt)
        {
            RequestForm requestForm = request.CreateRequestForm();
            Dictionary<string, object> form = requestForm.Form;
            string dataName = this.openConfig.DataName;
            object data = form[dataName];

            string dataJson = JsonUtil.ToJSONString(data);
          
            dataJson = HttpUtility.UrlEncode(dataJson, System.Text.Encoding.UTF8);

            form[dataName] = dataJson;
            form.Add(this.openConfig.AppKeyName, this.appKey); 
            // 添加秘钥
            string sign = SignUtil.CreateSign(form, this.secret);
            form.Add(this.openConfig.SignName, sign);

            Dictionary<string, string> header = this.buildHeader(jwt);

            string resp = this.doExecute(url, requestForm, header);

            return this.parseResponse<T>(resp, request);
        }

        /// <summary>
        /// 执行请求
        /// </summary>
        /// <param name="url">请求url</param>
        /// <param name="requestForm">请求内容</param>
        /// <param name="header">请求header</param>
        /// <returns>返回服务器响应内容</returns>
        protected virtual String doExecute(String url, RequestForm requestForm, Dictionary<string, string> header)
        {
            return openRequest.Request(this.url, requestForm, header);
        }

        /// <summary>
        /// 解析返回结果
        /// </summary>
        /// <typeparam name="T">返回的Response</typeparam>
        /// <param name="resp">服务器响应内容</param>
        /// <param name="request">请求Request</param>
        /// <returns>返回Response</returns>
        protected virtual T parseResponse<T>(String resp, BaseRequest<T> request) {
            return JsonUtil.ParseObject<T>(resp);
        }


        private Dictionary<string, string> buildHeader(string jwt)
        {
            Dictionary<string, string> header = new Dictionary<string, string>();
            header.Add(ACCEPT_LANGUAGE, this.openConfig.Locale);
            if (!string.IsNullOrEmpty(jwt))
            {
                header.Add(AUTHORIZATION, PREFIX_BEARER + jwt);
            }
            return header;
        }

    }
}

File: icsopenapi/easyopen-sdk/sdk-java/src/main/java/com/easyopen/sdk/client/OpenClient.java
package com.easyopen.sdk.client;

import com.easyopen.sdk.common.OpenConfig;
import com.easyopen.sdk.common.RequestForm;
import com.easyopen.sdk.request.BaseRequest;
import com.easyopen.sdk.response.BaseResponse;
import com.easyopen.sdk.util.JsonUtil;
import com.easyopen.sdk.util.SignUtil;
import com.easyopen.sdk.util.StringUtil;

import java.util.HashMap;
import java.util.Map;

/**
 * 请求客户端
 * @author tanghc
 */
public class OpenClient {
    private static final String ACCEPT_LANGUAGE = "Accept-Language";
    private static final String AUTHORIZATION = "Authorization";
    private static final String PREFIX_BEARER = "Bearer ";
    private static final OpenConfig DEFAULT_CONFIG = new OpenConfig();

    private String url;
    private String appKey;
    private String secret;

    private OpenConfig openConfig;
    private OpenRequest openRequest;

    public OpenClient(String url, String appKey, String secret) {
        this(url, appKey, secret, DEFAULT_CONFIG);
    }

    public OpenClient(String url, String appKey, String secret, OpenConfig openConfig) {
        if (openConfig == null) {
            throw new IllegalArgumentException("openConfig不能为null");
        }
        this.url = url;
        this.appKey = appKey;
        this.secret = secret;
        this.openConfig = openConfig;

        this.openRequest = new OpenRequest(openConfig);
    }

    /**
     * 请求接口
     * @param request 请求对象
     * @param <T> 返回对应的Response
     * @return 返回Response
     */
    public <T extends BaseResponse<?>> T execute(BaseRequest<T> request) {
        return this.execute(request, null);
    }

    /**
     * 请求接口
     * @param request 请求对象
     * @param jwt jwt
     * @param <T> 返回对应的Response
     * @return 返回Response
     */
    public <T extends BaseResponse<?>> T execute(BaseRequest<T> request, String jwt) {
        RequestForm requestForm = request.createRequestForm();
        // 表单数据
        Map<String, Object> form = requestForm.getForm();

        form.put(this.openConfig.getAppKeyName(), this.appKey);
        // 将data部分转成json并urlencode
        Object data = form.get(this.openConfig.getDataName());
        String dataJson = JsonUtil.toJSONString(data);
        dataJson = StringUtil.encodeUrl(dataJson);
        form.put(this.openConfig.getDataName(), dataJson);
        // 生成签名，并加入到form中
        String sign = SignUtil.createSign(form, this.secret);
        form.put(this.openConfig.getSignName(), sign);

        // 构建http请求header
        Map<String, String> header = this.buildHeader(jwt);

        String resp = doExecute(this.url, requestForm, header);

        return this.parseResponse(resp, request);
    }

    protected String doExecute(String url, RequestForm requestForm, Map<String, String> header) {
        return openRequest.request(this.url, requestForm, header);
    }

    protected <T extends BaseResponse<?>> T parseResponse(String resp, BaseRequest<T> request) {
        return JsonUtil.parseObject(resp, request.getResponseClass());
    }

    protected Map<String, String> buildHeader(String jwt) {
        Map<String, String> header = new HashMap<String, String>();
        header.put(ACCEPT_LANGUAGE, this.openConfig.getLocale());
        if (StringUtil.isNotEmpty(jwt)) {
            header.put(AUTHORIZATION, PREFIX_BEARER + jwt);
        }
        return header;
    }

    public OpenRequest getOpenRequest() {
        return openRequest;
    }

    public void setOpenRequest(OpenRequest openRequest) {
        this.openRequest = openRequest;
    }

    public String getAppKey() {
        return appKey;
    }

    public void setAppKey(String appKey) {
        this.appKey = appKey;
    }

    public OpenConfig getOpenConfig() {
        return openConfig;
    }

}
